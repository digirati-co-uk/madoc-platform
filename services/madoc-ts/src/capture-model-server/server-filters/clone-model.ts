import { generateId } from '../../frontend/shared/capture-models/helpers/generate-id';
import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel, Target } from '../../frontend/shared/capture-models/types/capture-model';

export function cloneModel(model: CaptureModel, { target }: { target?: Target[] }): CaptureModel {
  const id = model.id;
  // @todo handle cloning with revisions
  const idMap: Record<string, string> = {};
  traverseDocument(model.document, {
    visitEntity(entity) {
      const newId = generateId();
      idMap[entity.id] = newId;
      entity.id = newId;
      entity.revision = undefined;
    },
    visitField(field) {
      const newId = generateId();
      idMap[field.id] = newId;
      field.id = newId;
      field.revision = undefined;
      field.revises = undefined;
    },
    visitSelector(selector) {
      const newId = generateId();
      idMap[selector.id] = newId;
      selector.id = newId;
      selector.revisionId = undefined;
    },
  });

  // Apply the revises field.
  traverseDocument(model.document, {
    visitEntity(entity) {
      if (entity.revises) {
        entity.revises = idMap[entity.revises] || undefined;
      }
    },
    visitField(field) {
      if (field.revises) {
        field.revises = idMap[field.revises] || undefined;
      }
    },
    visitSelector(selector) {
      if (selector.revises) {
        selector.revises = idMap[selector.revises] || undefined;
      }
    },
  });

  model.id = generateId();
  model.revisions = [];
  // @todo contributors.
  // model.contributors = creator ? { [creator.id]: creator } : {};
  model.derivedFrom = id;
  model.target = target;

  return model;
}
