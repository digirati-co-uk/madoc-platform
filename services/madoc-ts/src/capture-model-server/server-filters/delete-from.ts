import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function deleteFrom(id: string, property: string, parent: CaptureModel['document']) {
  if ((parent as any).properties[property]) {
    (parent as any).properties[property] = (parent as any).properties[property].filter((e: any) => e.id !== id);
  }
}
