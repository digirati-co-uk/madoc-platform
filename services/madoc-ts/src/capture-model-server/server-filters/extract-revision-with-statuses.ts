import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function extractRevisionWithStatuses(model: CaptureModel, statuses: string[]): string[] {
  const toReturn = [];
  if (!model.revisions) {
    return [];
  }

  for (const revision of model.revisions) {
    if (revision.status) {
      if (statuses.indexOf(revision.status) !== -1) {
        toReturn.push(revision.id);
      }
    }
  }

  return toReturn;
}
