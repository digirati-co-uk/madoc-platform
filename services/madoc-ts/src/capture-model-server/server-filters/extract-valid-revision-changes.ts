import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { validateRevision } from '../../frontend/shared/capture-models/helpers/validate-revision';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { BaseField } from '../../frontend/shared/capture-models/types/field-types';
import { RevisionRequest } from '../../frontend/shared/capture-models/types/revision-request';
import { BaseSelector } from '../../frontend/shared/capture-models/types/selector-types';
import { getBefore } from './get-before';

export function extractValidRevisionChanges(
  captureModel: CaptureModel,
  req: RevisionRequest,
  {
    allowAnonymous,
    allowCanonicalChanges,
    allowCustomStructure,
  }: {
    allowAnonymous?: boolean;
    allowCanonicalChanges?: boolean;
    allowCustomStructure?: boolean;
  }
) {
  if (!req.document.immutable) {
    req.document.immutable = true;
  }

  const filteredDocument = validateRevision(req, captureModel, {
    allowAnonymous,
    allowCanonicalChanges,
    allowCustomStructure,
  });
  const allIds: string[] = [];
  const mutations: Record<
    string,
    Array<
      | {
          id: string;
          type: 'deletion';
        }
      | {
          id: string;
          type: 'entity';
          entity: CaptureModel['document'];
          term: string;
          shallow?: boolean;
          before: [string | null, number];
        }
      | {
          id: string;
          type: 'field';
          field: BaseField;
          term: string;
          before: [string | null, number];
        }
      | {
          id: string;
          type: 'selector';
          selector: BaseSelector;
          parentSelector?: BaseSelector;
        }
      | {
          id: string;
          type: 'selector-self';
          selector: BaseSelector;
        }
    >
  > = {};

  if (filteredDocument) {
    // @todo if `allowCanonicalChanges` is true, then this strategy probably will not work.
    //   When this is set we need to allow updating all properties of the included entities and fields
    //   But we can assume they might be incomplete (not all properties in the document for example).
    //   A better strategy might be to simply replace the values shallow-ly

    // 1st traverse the revision to get the fields that need added.
    traverseDocument(filteredDocument, {
      visitField(field, term, parent) {
        // Removing this immutability check. Not sure where it was required.
        if (allowCanonicalChanges || /*parent.immutable && */ field.revision === req.revision.id) {
          allIds.push(field.id);
          mutations[parent.id] = mutations[parent.id] ? mutations[parent.id] : [];
          mutations[parent.id].push({
            id: field.id,
            type: 'field',
            field,
            term,
            before: getBefore(field, term, parent),
          });
        }
      },
      visitEntity(entity, term, parent) {
        if (parent && term && (allowCanonicalChanges || (!entity.immutable && entity.revision === req.revision.id))) {
          allIds.push(entity.id);
          mutations[parent.id] = mutations[parent.id] ? mutations[parent.id] : [];
          mutations[parent.id].push({
            id: entity.id,
            type: 'entity',
            entity,
            term,
            shallow: allowCanonicalChanges,
            before: getBefore(entity, term, parent),
          });
        }
      },
      visitSelector(selector, parent, isRevision, parentSelector) {
        if (allowCanonicalChanges) {
          allIds.push(selector.id);
          mutations[selector.id] = mutations[selector.id] ? mutations[selector.id] : [];
          mutations[selector.id].push({
            id: selector.id,
            type: 'selector-self',
            selector,
          });
        } else if (isRevision && parentSelector && selector.revisionId && selector.revisionId === req.revision.id) {
          allIds.push(selector.id);
          mutations[parentSelector.id] = mutations[parentSelector.id] ? mutations[parentSelector.id] : [];
          mutations[parentSelector.id].push({ id: selector.id, type: 'selector', selector, parentSelector });
        }
      },
    });
  }

  // Optionally traverse to find valid fields to be removed.
  const requestedToDelete = req.revision.deletedFields || [];
  const validToDelete: string[] = [];

  traverseDocument(captureModel.document, {
    visitEntity(entity) {
      if (!entity.immutable && requestedToDelete.indexOf(entity.id) !== -1) {
        validToDelete.push(entity.id);
      }
      if (allIds.indexOf(entity.id) === -1 && entity.revision === req.revision.id) {
        // This is a deletion
        mutations[entity.id] = [{ type: 'deletion', id: entity.id }];
      }
    },
    visitField(field) {
      if (!field.immutable && requestedToDelete.indexOf(field.id) !== -1) {
        validToDelete.push(field.id);
      }
      if (allIds.indexOf(field.id) === -1 && field.revision === req.revision.id) {
        // This is a deletion
        mutations[field.id] = [{ type: 'deletion', id: field.id }];
      }
    },
    visitSelector(selector, _, isRevision) {
      if (selector && isRevision && allIds.indexOf(selector.id) === -1 && selector.revisionId === req.revision.id) {
        // This is a deletion
        mutations[selector.id] = [{ type: 'deletion', id: selector.id }];
      }
    },
  });

  req.revision.deletedFields = validToDelete;

  return {
    mutations,
    deletions: validToDelete,
  };
}
