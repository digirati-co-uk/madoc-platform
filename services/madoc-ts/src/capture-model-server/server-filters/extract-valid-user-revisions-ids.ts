import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';

export function extractValidUserRevisionsIds(model: CaptureModel, userId: number) {
  const userUrn = `urn:madoc:user:${userId}`;
  const returnMap = {
    myRevisions: [] as string[],
    approvedRevisions: [] as string[],
    excludedRevisions: [] as string[],
  };

  if (!model.revisions) {
    return returnMap;
  }

  for (const revision of model.revisions) {
    let visible = false;
    if (revision.approved) {
      visible = true;
      returnMap.approvedRevisions.push(revision.id);
    }
    if (revision.authors && revision.authors.indexOf(userUrn) !== -1) {
      visible = true;
      returnMap.myRevisions.push(revision.id);
    }

    if (!visible) {
      returnMap.excludedRevisions.push(revision.id);
    }
  }

  return returnMap;
}
