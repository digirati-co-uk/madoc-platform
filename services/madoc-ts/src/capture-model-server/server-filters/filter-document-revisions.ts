import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { deleteFrom } from './delete-from';

export function filterDocumentRevisions(
  document: CaptureModel['document'],
  excludeRevisions: string[],
  onlyRevisionFields?: boolean
): CaptureModel['document'] {
  traverseDocument(document, {
    beforeVisitEntity(entity, property, parent) {
      if (property && parent) {
        if (onlyRevisionFields && !entity.revision) {
          deleteFrom(entity.id, property, parent);
        }

        if (entity.revision && excludeRevisions.indexOf(entity.revision) !== -1) {
          deleteFrom(entity.id, property, parent);
        }
      }
    },
    visitField(field, property, parent) {
      if (onlyRevisionFields && !field.revision) {
        deleteFrom(field.id, property, parent);
      }

      if (field.revision && property && parent) {
        if (excludeRevisions.indexOf(field.revision) !== -1) {
          deleteFrom(field.id, property, parent);
        }
      }
    },
    visitSelector(selector, parent, isRevision, parentSelector) {
      if (selector.revisionId && isRevision && parentSelector && parentSelector.revisedBy) {
        parentSelector.revisedBy = parentSelector.revisedBy.filter(s => {
          if (!s.revisionId) {
            return s;
          }
          return excludeRevisions.indexOf(s.revisionId) === -1;
        });
      }
    },
  });

  return document;
}
