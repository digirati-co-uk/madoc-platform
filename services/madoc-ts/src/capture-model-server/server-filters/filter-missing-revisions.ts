import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { deleteFrom } from './delete-from';

export function filterMissingRevisions(captureModel: CaptureModel) {
  const allRevisions = (captureModel.revisions || []).map(r => r.id);

  traverseDocument(captureModel.document, {
    beforeVisitEntity(entity, property, parent) {
      if (property && parent && entity.revision && allRevisions.indexOf(entity.revision) === -1) {
        deleteFrom(entity.id, property, parent);
      }
    },
    visitField(field, property, parent) {
      if (property && parent && field.revision && allRevisions.indexOf(field.revision) === -1) {
        deleteFrom(field.id, property, parent);
      }
    },
    visitSelector(selector, parent, isRevision, parentSelector) {
      if (selector.revisionId && isRevision && parentSelector && parentSelector.revisedBy) {
        parentSelector.revisedBy = parentSelector.revisedBy.filter(s => {
          if (!s.revisionId) {
            return s;
          }
          return allRevisions.indexOf(s.revisionId) !== -1;
        });
      }
    },
  });

  return captureModel;
}
