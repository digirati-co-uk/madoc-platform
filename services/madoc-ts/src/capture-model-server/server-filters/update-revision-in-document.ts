import { traverseDocument } from '../../frontend/shared/capture-models/helpers/traverse-document';
import { CaptureModel } from '../../frontend/shared/capture-models/types/capture-model';
import { RevisionRequest } from '../../frontend/shared/capture-models/types/revision-request';
import { deleteFrom } from './delete-from';
import { extractValidRevisionChanges } from './extract-valid-revision-changes';

export function updateRevisionInDocument(
  captureModel: CaptureModel,
  req: RevisionRequest,
  options: {
    allowAnonymous?: boolean;
    allowCanonicalChanges?: boolean;
    allowCustomStructure?: boolean;
  } = {}
) {
  const { mutations } = extractValidRevisionChanges(captureModel, req, {
    allowAnonymous: options.allowAnonymous,
    allowCustomStructure: options.allowCustomStructure,
    allowCanonicalChanges: false,
  });

  if (Object.keys(mutations).length) {
    // Apply document mutations.
    traverseDocument(captureModel.document, {
      visitSelector(selector) {
        const mutationsToApply = mutations[selector.id];
        if (mutationsToApply && mutationsToApply.length) {
          for (const mutation of mutationsToApply) {
            if (mutation.type === 'deletion') {
              // @todo delete selectors (not currently available in UI)
            }

            if (mutation.type === 'selector') {
              selector.revisedBy = selector.revisedBy ? selector.revisedBy : [];
              const existingSelector = selector.revisedBy.findIndex(s => s.id === mutation.id);
              if (existingSelector !== -1) {
                selector.revisedBy[existingSelector] = mutation.selector;
              } else {
                selector.revisedBy.push(mutation.selector);
              }
            }
          }
        }
      },

      visitField(field, property, parent) {
        const mutationsToApply = mutations[field.id];
        if (mutationsToApply && mutationsToApply.length) {
          for (const mutation of mutationsToApply) {
            // Deletion of an item that matches the revision.
            if (mutation.type === 'deletion') {
              if (parent && property) {
                deleteFrom(field.id, property, parent);
              }
              continue;
            }
          }
        }
      },

      visitEntity(entity, property, parent) {
        // Is there a change to be applied
        const mutationsToApply = mutations[entity.id];
        if (mutationsToApply && mutationsToApply.length) {
          for (const mutation of mutationsToApply) {
            // Deletion of an item that matches the revision.
            if (mutation.type === 'deletion') {
              if (parent && property) {
                deleteFrom(entity.id, property, parent);
              }
              continue;
            }

            // Selectors on their own.
            if (mutation.type === 'selector') {
              if (entity.selector) {
                entity.selector.revisedBy = entity.selector.revisedBy ? entity.selector.revisedBy : [];
                const existingSelector = entity.selector.revisedBy.findIndex(s => s.id === mutation.id);
                if (existingSelector !== -1) {
                  entity.selector.revisedBy[existingSelector] = mutation.selector;
                } else {
                  entity.selector.revisedBy.push(mutation.selector);
                }
              }
              continue;
            }

            if (mutation.type === 'selector-self') {
              const revised = entity.selector?.revisedBy;
              entity.selector = mutation.selector;
              // Don't allow updating the revised property directly here. Avoid messing up other peoples revisions.
              if (revised) {
                entity.selector.revisedBy = revised;
              }
              continue;
            }

            // Properties.
            const propertyList = entity.properties[mutation.term as any];
            if (!propertyList) {
              // Maybe configuration?
              entity.properties[mutation.term as any] = [];
            }

            const existing = (entity.properties[mutation.term as any] as any[]).findIndex(
              (r: any) => r.id === mutation.id
            );
            if (existing !== -1) {
              // Replace with a null value placeholder (filtered out further down) to preserve the index.
              (entity.properties[mutation.term as any][existing] as any) = null;
            }
            let toPush = mutation.type === 'entity' ? mutation.entity : mutation.field;

            // This performs a shallow merge, since the other properties should already be applied
            // in a previous step.
            if (mutation.type === 'entity' && mutation.shallow && existing !== -1) {
              const existingProperties = entity.properties[mutation.term as any][existing];
              if (existingProperties) {
                toPush = {
                  ...(toPush as any),
                  selector: existingProperties.selector,
                  properties: (existingProperties as any).properties,
                } as any;
              }
            }

            // then we are doing an insertion.
            const [beforeId] = mutation.before;
            const beforeIndex = (entity.properties[mutation.term] as any[]).findIndex(r => r && r.id === beforeId);
            if (beforeIndex !== -1) {
              (entity.properties[mutation.term as any] as any[]).splice(beforeIndex + 1, 0, toPush);
            } else {
              (entity.properties[mutation.term as any] as any[]).push(toPush);
            }

            // Filter our placeholder used to keep track of the correct beforeIndex.
            (entity.properties[mutation.term as any] as any[]) = (entity.properties[
              mutation.term as any
            ] as any[]).filter(r => r !== null);
          }
        }
      },
    });
  }
}
