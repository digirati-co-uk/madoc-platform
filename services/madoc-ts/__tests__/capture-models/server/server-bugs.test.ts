import { updateRevisionInDocument } from '../../../src/capture-model-server/server-filters/update-revision-in-document';

describe('Capture model server bugs', () => {
  test('Japans project - should not delete original fields', () => {
    const fixture = {
      captureModel: {
        id: 'd3944d3b-4aaf-4eac-bbbb-7a978084866b',
        structure: {
          id: '925d7b59-ec4f-43b0-bc67-f6cc039de112',
          label: 'klassiek-japans',
          type: 'choice',
          items: [
            {
              id: 'df80021e-a6e9-4b92-8276-7f86ea67eec7',
              type: 'model',
              label: 'Default',
              fields: [['vertaling en transcriptie', ['vertaling', 'transcriptie']]],
            },
          ],
        },
        document: {
          id: '78231d5f-0ac9-46fb-8914-07c6cd0bdb82',
          type: 'entity',
          label: 'klassiek-japans',
          properties: {
            'vertaling en transcriptie': [
              {
                id: '4b1ef70a-fdf0-4ecd-9a1e-fe645765f9c4',
                type: 'entity',
                label: 'vertaling en transcriptie',
                selector: {
                  id: '4bb0cc03-56f5-4f86-91a6-a753ccf023c4',
                  type: 'box-selector',
                  state: null,
                  required: true,
                },
                properties: {
                  vertaling: [
                    { id: '8e799acb-90c9-4852-9e8c-ba7903c43fa7', type: 'text-field', label: 'vertaling', value: '' },
                  ],
                  transcriptie: [
                    {
                      id: '3ebbe019-5448-48f9-a92a-459aa59d6219',
                      type: 'text-field',
                      label: 'transcriptie',
                      value: '',
                    },
                  ],
                },
                allowMultiple: true,
              },
            ],
          },
        },
        revisions: [],
        derivedFrom: '19abcfe1-6303-48c7-9cd2-356302bb9d47',
        target: [
          { id: 'urn:madoc:manifest:173', type: 'Manifest' },
          { id: 'urn:madoc:canvas:286', type: 'Canvas' },
        ],
        contributors: {},
      },
      structure: {
        id: 'df80021e-a6e9-4b92-8276-7f86ea67eec7',
        type: 'model',
        label: 'Default',
        fields: [['vertaling en transcriptie', ['vertaling', 'transcriptie']]],
      },
      filterEmpty: true,
    } as any;

    const request = {
      captureModelId: 'd3944d3b-4aaf-4eac-bbbb-7a978084866b',
      revision: {
        id: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
        fields: [['vertaling en transcriptie', ['vertaling', 'transcriptie']]],
        approved: false,
        structureId: 'df80021e-a6e9-4b92-8276-7f86ea67eec7',
        label: 'Default',
        revises: 'df80021e-a6e9-4b92-8276-7f86ea67eec7',
      },
      document: {
        id: '78231d5f-0ac9-46fb-8914-07c6cd0bdb82',
        type: 'entity',
        label: 'klassiek-japans',
        properties: {
          'vertaling en transcriptie': [
            {
              id: '4b1ef70a-fdf0-4ecd-9a1e-fe645765f9c4',
              type: 'entity',
              label: 'vertaling en transcriptie',
              selector: {
                id: '4bb0cc03-56f5-4f86-91a6-a753ccf023c4',
                type: 'box-selector',
                state: null,
                required: true,
                revisedBy: [
                  {
                    id: '2e41606c-d35a-41da-9700-b0d434f2084d',
                    type: 'box-selector',
                    state: { x: 716, y: 1594, width: 579, height: 557 },
                    required: true,
                    revisionId: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
                    revises: '4bb0cc03-56f5-4f86-91a6-a753ccf023c4',
                  },
                ],
              },
              properties: {
                vertaling: [
                  {
                    id: '59cd2ead-25fd-4384-bbb8-d435fd02d1d0',
                    type: 'text-field',
                    label: 'vertaling',
                    value: 'a',
                    revision: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
                    revises: '8e799acb-90c9-4852-9e8c-ba7903c43fa7',
                  },
                ],
                transcriptie: [
                  {
                    id: 'd544a7d9-3c6c-4d62-aaaa-548330ae9da3',
                    type: 'text-field',
                    label: 'transcriptie',
                    value: 'a',
                    revision: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
                    revises: '3ebbe019-5448-48f9-a92a-459aa59d6219',
                  },
                ],
              },
              allowMultiple: true,
              immutable: true,
            },
            {
              id: '76abd494-2ead-4a81-995f-63e730d25b2f',
              type: 'entity',
              label: 'vertaling en transcriptie',
              selector: {
                id: '931e9391-f8b3-4453-93b5-9b879285c65d',
                type: 'box-selector',
                state: { x: 1488, y: 1344, width: 216, height: 239 },
                required: true,
              },
              allowMultiple: true,
              immutable: false,
              properties: {
                vertaling: [
                  {
                    id: '6df72c26-4e94-49cd-ae2c-fa15b76ab7ae',
                    type: 'text-field',
                    label: 'vertaling',
                    value: 'b',
                    revision: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
                  },
                ],
                transcriptie: [
                  {
                    id: '255af78e-6fef-4cd6-b65d-973584c73efa',
                    type: 'text-field',
                    label: 'transcriptie',
                    value: 'b',
                    revision: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
                  },
                ],
              },
              revision: 'f284c0e6-1b55-4268-af1a-cba8c2056946',
            },
          ],
        },
        immutable: true,
      },
      source: 'structure',
    } as any;

    updateRevisionInDocument(fixture.captureModel, request, {
      allowAnonymous: true,
      allowCanonicalChanges: true,
    });

    expect(fixture.captureModel).toMatchSnapshot();

    //
  });
});
