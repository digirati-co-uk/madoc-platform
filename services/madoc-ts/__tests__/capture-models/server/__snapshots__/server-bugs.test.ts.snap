// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capture model server bugs Japans project - should not delete original fields 1`] = `
Object {
  "contributors": Object {},
  "derivedFrom": "19abcfe1-6303-48c7-9cd2-356302bb9d47",
  "document": Object {
    "id": "78231d5f-0ac9-46fb-8914-07c6cd0bdb82",
    "label": "klassiek-japans",
    "properties": Object {
      "vertaling en transcriptie": Array [
        Object {
          "allowMultiple": true,
          "id": "4b1ef70a-fdf0-4ecd-9a1e-fe645765f9c4",
          "label": "vertaling en transcriptie",
          "properties": Object {
            "transcriptie": Array [
              Object {
                "id": "3ebbe019-5448-48f9-a92a-459aa59d6219",
                "label": "transcriptie",
                "type": "text-field",
                "value": "",
              },
              Object {
                "id": "d544a7d9-3c6c-4d62-aaaa-548330ae9da3",
                "immutable": true,
                "label": "transcriptie",
                "revises": "3ebbe019-5448-48f9-a92a-459aa59d6219",
                "revision": "f284c0e6-1b55-4268-af1a-cba8c2056946",
                "type": "text-field",
                "value": "a",
              },
            ],
            "vertaling": Array [
              Object {
                "id": "8e799acb-90c9-4852-9e8c-ba7903c43fa7",
                "label": "vertaling",
                "type": "text-field",
                "value": "",
              },
              Object {
                "id": "59cd2ead-25fd-4384-bbb8-d435fd02d1d0",
                "immutable": true,
                "label": "vertaling",
                "revises": "8e799acb-90c9-4852-9e8c-ba7903c43fa7",
                "revision": "f284c0e6-1b55-4268-af1a-cba8c2056946",
                "type": "text-field",
                "value": "a",
              },
            ],
          },
          "selector": Object {
            "id": "4bb0cc03-56f5-4f86-91a6-a753ccf023c4",
            "required": true,
            "revisedBy": Array [
              Object {
                "id": "2e41606c-d35a-41da-9700-b0d434f2084d",
                "required": true,
                "revises": "4bb0cc03-56f5-4f86-91a6-a753ccf023c4",
                "revisionId": "f284c0e6-1b55-4268-af1a-cba8c2056946",
                "state": Object {
                  "height": 557,
                  "width": 579,
                  "x": 716,
                  "y": 1594,
                },
                "type": "box-selector",
              },
            ],
            "state": null,
            "type": "box-selector",
          },
          "type": "entity",
        },
        Object {
          "allowMultiple": true,
          "id": "76abd494-2ead-4a81-995f-63e730d25b2f",
          "immutable": false,
          "label": "vertaling en transcriptie",
          "properties": Object {
            "transcriptie": Array [
              Object {
                "id": "255af78e-6fef-4cd6-b65d-973584c73efa",
                "label": "transcriptie",
                "revision": "f284c0e6-1b55-4268-af1a-cba8c2056946",
                "type": "text-field",
                "value": "b",
              },
            ],
            "vertaling": Array [
              Object {
                "id": "6df72c26-4e94-49cd-ae2c-fa15b76ab7ae",
                "label": "vertaling",
                "revision": "f284c0e6-1b55-4268-af1a-cba8c2056946",
                "type": "text-field",
                "value": "b",
              },
            ],
          },
          "revision": "f284c0e6-1b55-4268-af1a-cba8c2056946",
          "selector": Object {
            "id": "931e9391-f8b3-4453-93b5-9b879285c65d",
            "required": true,
            "state": Object {
              "height": 239,
              "width": 216,
              "x": 1488,
              "y": 1344,
            },
            "type": "box-selector",
          },
          "type": "entity",
        },
      ],
    },
    "type": "entity",
  },
  "id": "d3944d3b-4aaf-4eac-bbbb-7a978084866b",
  "revisions": Array [],
  "structure": Object {
    "id": "925d7b59-ec4f-43b0-bc67-f6cc039de112",
    "items": Array [
      Object {
        "fields": Array [
          Array [
            "vertaling en transcriptie",
            Array [
              "vertaling",
              "transcriptie",
            ],
          ],
        ],
        "id": "df80021e-a6e9-4b92-8276-7f86ea67eec7",
        "label": "Default",
        "type": "model",
      },
    ],
    "label": "klassiek-japans",
    "type": "choice",
  },
  "target": Array [
    Object {
      "id": "urn:madoc:manifest:173",
      "type": "Manifest",
    },
    Object {
      "id": "urn:madoc:canvas:286",
      "type": "Canvas",
    },
  ],
}
`;
